haystack/agents/memory/conversation_summary_memory.py

```diff
@@ -0,0 +1,87 @@
+from typing import Optional, Union, Dict, Any, List
+
+from haystack.agents.memory import ConversationMemory
+from haystack.nodes import PromptTemplate, PromptNode
+
+
+class ConversationSummaryMemory(ConversationMemory):
+    """
+    A memory class that stores conversation history and periodically generates summaries.
+    """
+
+    def __init__(
+        self,
+        pn: PromptNode,
+        prompt_template: Optional[Union[str, PromptTemplate]] = None,
```

Comment:
Didn't we agree to remove `prompt_template` and rely in the one set in `PromptNode`?

Also let's rename `pn`, `prompt_node` would be fine.

Reply:
I think we agreed on standardizing approach for all components where prompt node and templates are user. Lets do that in a separate PRs across the board

Reply:
> Didn't we agree to remove `prompt_template` and rely in the one set in `PromptNode`?
> 
> 
> 
> Also let's rename `pn`, `prompt_node` would be fine.

I'll rename, my bad but lets do standardization for all in one PR

Reply:
I think it's fine doing it in a separate PR for other components but since this is new let's do it now since we're at it.

Reply:
Ok, I just checked the algorithm; it is what I believe to be the one that makes the most sense. If the user provides a template - we use it. If no template is provided, then the default template of the prompt node is used, and if that one is not specified, then we use the one from the default templates - "conversation-summary". WDYT?

Reply:
Let's keep it simple. The user must pass a `PromptNode` and we use whatever template is set for that node.

Reply:
I do agree that this is certainly easier to understand for a beginner, but I don't agree that this is for the best. 

With this optional template we hide the complexity and a big strong requirement to use a `ConversationSummaryMemory` correctly. And that is use the right prompt template to handle memory.

The user must know that to use memory correctly it must use the right prompt and I think that must be explicit, and we do that by teaching the user that memory works only with a correctly configured `PromptNode`.

This snippet here must not work:

```
prompt_node = PromptNode("gpt-3.5-turbo", api_key=os.environ.get("OPENAI_API_KEY"), max_length=256) 
agent = ConversationalAgent(prompt_node, memory=ChatSummaryMemory(prompt_node))
```

This instead must work:

```
prompt_node = PromptNode("gpt-3.5-turbo", template="conversational-summary", api_key=os.environ.get("OPENAI_API_KEY"), max_length=256) 
agent = ConversationalAgent(prompt_node, memory=ChatSummaryMemory(prompt_node))
```

Also if we give the possility to set a custom template in here it gives the chance for users to create this kind of code that is really confusing and doesn't make much sense:

```
prompt_node = PromptNode("gpt-3.5-turbo", template="zero-shot-react", api_key=os.environ.get("OPENAI_API_KEY"), max_length=256) 
agent = ConversationalAgent(prompt_node, memory=ChatSummaryMemory(prompt_node), template="conversational-summary")
```

In the end the `PromptNode` must be the single source of truth of which template will be used by the memory. 

All the other functions and component that accept both a `PromptNode` and a template must be gone and get the same treatment. 

If we want to let users change the prompt template on the fly we might as well decouple `PromptNode` and `PromptTemplate` and force users to specify the template each time they want to make use of a `PromptNode`. But I think that's an even worse way of using it.